Werkgroep API design rules/naamgeving.md

```diff
@@ -0,0 +1,56 @@
+    - Zeker wanneer complexere vormen van filtering worden toegestaan, bijvoorbeeld `/collectie?filter[status]=open` in plaats van `/collectie?status=open`. 
+    - Tevens zullen conflicten tussen *technische* (bv. `sort`) en functionele (bv `status` in bovenstaand voorbeeld) beperkt blijven omdat technische request parameters Engelstalig zijn en functionele over het algemeen Nederlandstalig.
+- Verdere uitwerking van het gebruik van de technische request parameters voor filteren en sorteren is onderdeel van de werkgroep Filters en Sorteringen.
+
+## Enumeraties
+Het voorstel voor enumeraties is:
+
+- (MUST) Gebruik technische waarden in enumeraties die los staan van presentatie van deze waarden.
+- In het geval dat het domeinmodel presentatie-specifieke waarden bevat (bijvoorbeeld BAG objectstatusen), gebruik daarvoor bij voorkeur GEEN enumeraties maar strings; documenteer wel de mogelijke waarden goed in de API documentatie.
+    - Nadelen van het gebruik van presentatie-specifieke waarden als enums zijn:
+        - het is voor Clients lastig om fouten te maken wanneer een waarde moet worden opgegeven of aangepast middels een `POST`, `PUT` of `PATCH` operatie.
+        - wanneer een enum waarde moet worden aangepast omwille de behoefte om de presentatie te wijzigen, betekent dit een breaking API change.
+        - in het geval van meertaligheid moeten API Clients verschillende logica hanteren om enum waarden uit te lezen of te wijzigen, afhankelijk van de geselecteerde taal.
+    - Beter is om naast de presentatie-specifieke waarde ook een enum aan te bieden en de presentatie-specifieke waarde op basis van de (technische) enum waarde te laten berekenen door de API.
+        - Als voorbeeld voor de BAG object statusen kan bijvoorbeeld worden gekozen voor een veld pand.status.naam, welke een string is met waarden als `Bouw gestart` en een veld pand.status.code, welke een enum is met waarden als `BOUW_GESTART`. Wanneer de enum door de Client wordt gewijzigd, berekent de API een nieuwe waarde voor de `pand.status.naam`. Dit maakt het tevens gemakkelijk om meertaligheid toe te passen, waarbij `pand.status.naam` afhankelijk van de geselecteerde taal wordt gepresenteerd.
+- (SHOULD) Gebruik waar mogelijk UPPER_CAMEL_CASE, URI's voor enumeraties.
+
+## Niet uitgewerkt:
+- Het filteren en zoeken op sub-objecten in een resource (bv adres.straat om te zoeken op het veld straat in het object adres) --> onderdeel van werkgroep Filters & Sorteringen.
+- Identifiers voor relaties naar andere datasets, in het object of in HAL envelop (bijvoorbeeld nummeraanduiding van de BAG).
+- Het voorstel is om NIET te standaardiseren op OAS componenten, omdat dat slechts een technische implementatie is en verplichte standaardisatie problemen op kan leveren bij bijvoorbeeld OAS generatoren, frameworks etc. Hetgeen waar we op willen standaardiseren is het API ontwerp, dus hetgeen de OAS specificatie oplevert.
```

Comment:
ik ben het hier niet helemaal mee eens. Je kan niet alle regels (zoals casing) afdwingen, omdat OAS generatoren dan problemen kunnen krijgen. Maar je kan wel regels en adviezen geven die zorgen dat de andere richting, code generatoren nette code opleveren. Dus geen diacrieten of leestekens in componentnamen. En verwachten dat ten minste een consequente naamgeving wordt toegepast.

Ook bij andere typen gegevens, zoals properties kan je in de problemen komen met standaardiseren. Wanneer je bijvoorbeeld een externe standaard (her)gebruikt (zoals geoJSON of HAL) kan het ook zijn dat daarin dingen zitten die niet aan je standaard voldoen. Bijvoorbeeld enumeraties in geoJSON zijn niet uppercase:
```
    pointGeoJSON:
      required:
      - coordinates
      - type
      type: object
      properties:
        type:
          type: string
          enum:
          - Point
```

Wellicht moet je daar (in algemene zin) iets over zeggen. Dat deze standaarden niet hergebruik van bredere nationale of internationale standaarden mag voorkomen of verbieden.

Reply:
@joostfarla @jasperroes @fterpstra wellicht goed om iets met deze opmerking van @fsamwel te doen bij volgende redactionele slagen: wij beschrijven een aantal ontwerpprincipes voor API's, maar in sommige gevallen worden naast de API Design Rules ook andere standaarden toegepast (bijvoorbeeld GeoJSON zoals @fsamwel hierboven aangeeft). weke andere en soms conflicterende principes meebrengen. In een aantal gevallen denk ik dat andere (internationale) standaarden "voorgaan".
Het lijkt mij nuttig om dit 1x generiek op te schrijven, in plaats van bij elk principe "Gebruik spinal-case bij path segmenten, tenzij een andere internationale standaard gebruikt wordt die een andere casing conventie voorschrijft" (oid).

Reply:
@frankvanes @joostfarla @jasperroes Terecht punt en de insteek van de API Design Rules is altijd geweest dat je ze toepast bij gebrek aan bestaande verplichte (internationale) standaard. Dit staat nu niet heel expliciet in het document.
Klopt het overigens dat ik nergens meer een aparte extensie naamgeving kan terugvinden? ik kijk door het pull request heen en zie extra principes toegevoegd worden in JSON en URIs maar geen naamgeving extensie anders dan bovenstaande tekst dat de aparte naamgeving.md is weggehaald.

Reply:
Excuses voor mijn (veel te) late reactie. Het klopt dat naamgeving.md is weggehaald. Dit heb ik gedaan omdat ik bij het toevoegen van extra principes review commentaren heb meegenomen en deze niet altijd goed door kon voeren in naamgeving.md. Nadat ik vervolgens review commentaren kreeg op verouderde teksten in naamgeving.md besloot ik dat het beter is duidelijk te maken dat de informatie in naamgeving.md niet meer accuraat is.